src/uploadAssets.js

```diff
@@ -214,24 +214,32 @@ function getUploadFailures(assetRefMap, assetIds) {
 function setAssetReferences(assetRefMap, assetIds, options) {
   const {client, tag} = options
   const lookup = assetRefMap.values()
-  const patchTasks = assetIds.reduce((tasks, assetId) => {
+
+  // Collects patch tasks per document to avoid patching the same document multiple times
+  const patchTasksPerDoc = assetIds.reduce((tasks, assetId) => {
     const documents = lookup.next().value
     if (typeof assetId !== 'string') {
       return tasks
     }
 
-    return tasks.concat(
-      documents.map(({documentId, path}) => ({
-        documentId,
-        path,
-        assetId,
-      })),
-    )
-  }, [])
+    documents.forEach(({documentId, path}) => {
+      tasks[documentId] = tasks[documentId] || []
+      tasks[documentId].push({path, assetId})
+    })
+    return tasks
+  }, {})
+
+  const patchTasks = []
+  Object.keys(patchTasksPerDoc).forEach((documentId) => {
+    patchTasks.push({
+      documentId,
+      tasks: patchTasksPerDoc[documentId],
+    })
+  })
 
-  // We now have an array of simple tasks, each containing:
-  // {documentId, path, assetId}
-  // Instead of doing a single mutation per asset, let's batch them up
+  // We now have an array of tasks per document, each containing:
+  // {documentId: string, tasks: [{path, assetId}]}
+  // Instead of doing a single mutation per document, let's batch them up
```

Comment:
This may end up dong significantly more patch operations per batch.  That may not be a real concern, but if it is it shouldn't be too hard to update this batching logic to create batches roughly sized to the total number of tasks rather than the number of documents 🤷

Reply:
wdym?

Reply:
Ive updated it, given this:
* document A - 300 assets
* document B - 500 assets
* document C - 500 assets

This will generate 2 mutations:
* First with 2 patches to document A  and document B
* Second with 1 patch to document C